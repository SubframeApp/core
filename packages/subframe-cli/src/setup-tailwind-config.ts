import path from "path"
import prompts from "prompts"
import { printNode, Project, SourceFile, ts } from "ts-morph"
import { abortOnState } from "./sync-helpers"
import { getExpressionForDefaultExport, isDefaultExport } from "./transforms/is-export"
import {
  hasSubframeContentGlob,
  hasSubframeRequire,
  isContentProperty,
  isPresetsProperty,
  makeSubframeContentGlob,
  makeSubframeRequire,
} from "./transforms/tailwind"

function printManualTailwindSteps(cwd: string, subframeDirPath: string, prependText: string) {
  const subframePresetRequire = printNode(makeSubframeRequire(cwd, subframeDirPath))
  const subframeContentGlob = makeSubframeContentGlob(cwd, subframeDirPath)

  const warningMessage = `${prependText}
1. Append the following to the presets array, or create if it doesn't exist: ${subframePresetRequire}
2. Append the following to the content array: ${subframeContentGlob}

Your end results should look something like this:

module.exports = {
  presets: [${subframePresetRequire}], // added by Subframe
  content: [
    // already existing content
    "${subframeContentGlob}", // added by Subframe
  ],
  // everything else
};
`

  // makes the warning message cyan via https://stackoverflow.com/questions/9781218/how-to-change-node-jss-console-font-color
  console.log("\x1b[36m%s\x1b[0m", warningMessage)
}

export async function setupTailwindConfig(cwd: string, subframeDirPath: string) {
  const subframePresetRequireAST = makeSubframeRequire(cwd, subframeDirPath)
  const subframeContentGlob = makeSubframeContentGlob(cwd, subframeDirPath)

  const project = new Project({ compilerOptions: { allowJs: true } })

  project.addSourceFileAtPathIfExists(path.join(cwd, "tailwind.config.js"))

  const tailwindConfig = project.getSourceFile(path.join(cwd, "tailwind.config.js"))

  /** no Tailwind config, let's skip this step then */
  if (!tailwindConfig) {
    printManualTailwindSteps(
      cwd,
      subframeDirPath,
      "Subframe could not find a tailwind.config.js file. If you use a .ts file, you'll need to configure it manually:",
    )
    return
  }

  /** config before transformations */
  const initialText = tailwindConfig.print()

  // na√Øve heuristics to check that the
  // subframe preset and content glob have
  // been properly set up
  if (
    initialText.indexOf(
      // example: require("./<path-to-subframe>/tailwind.config.js")
      printNode(subframePresetRequireAST),
    ) !== -1 &&
    initialText.indexOf(
      // example: "./<path-to-subframe>/**/*.{tsx,ts,js,jsx}"
      subframeContentGlob,
    ) !== -1
  ) {
    return
  }

  const response = await prompts({
    type: "confirm",
    name: "updateTailwindConfig",
    initial: true,
    message: "Do you want Subframe to configure your Tailwind config?",
    onState: abortOnState,
  })
  if (!response.updateTailwindConfig) {
    return
  }

  transformTailwindConfigFile(tailwindConfig, cwd, subframeDirPath)

  await tailwindConfig.save()
}

// helper function that's exported for testing
// NOTE: mutates tailwindConfig file in place
export function transformTailwindConfigFile(tailwindConfig: SourceFile, cwd: string, subframeDirPath: string) {
  const subframePresetRequireAST = makeSubframeRequire(cwd, subframeDirPath)
  const subframeContentGlob = makeSubframeContentGlob(cwd, subframeDirPath)

  const initialText = tailwindConfig.print()

  tailwindConfig.transform((traversal) => {
    const node = traversal.visitChildren()

    if (isDefaultExport(node)) {
      const exportedExpression = getExpressionForDefaultExport(node)

      if (
        // assert that the default export is an object literal:
        !ts.isObjectLiteralExpression(exportedExpression)
      ) {
        return node
      }

      const properties = exportedExpression.properties
        /** transform the content and presets properties */
        .map((property) => {
          if (isContentProperty(property) && !hasSubframeContentGlob(property.initializer, cwd, subframeDirPath)) {
            return traversal.factory.updatePropertyAssignment(
              property,
              property.name,
              traversal.factory.createArrayLiteralExpression([
                ...property.initializer.elements,
                traversal.factory.createStringLiteral(subframeContentGlob),
              ]),
            )
          }

          if (isPresetsProperty(property) && !hasSubframeRequire(property.initializer, cwd, subframeDirPath)) {
            return traversal.factory.updatePropertyAssignment(
              property,
              property.name,
              traversal.factory.createArrayLiteralExpression([
                ...property.initializer.elements,
                subframePresetRequireAST,
              ]),
            )
          }

          return property
        })

      // if there's no content property, add it
      if (properties.findIndex(isContentProperty) === -1) {
        properties.push(
          traversal.factory.createPropertyAssignment(
            traversal.factory.createIdentifier("content"),
            traversal.factory.createArrayLiteralExpression([
              traversal.factory.createStringLiteral(subframeContentGlob),
            ]),
          ),
        )
      }

      // if there's no presets property, add it
      if (properties.findIndex(isPresetsProperty) === -1) {
        properties.push(
          traversal.factory.createPropertyAssignment(
            traversal.factory.createIdentifier("presets"),
            traversal.factory.createArrayLiteralExpression([subframePresetRequireAST]),
          ),
        )
      }

      if (ts.isExportAssignment(node)) {
        return traversal.factory.updateExportAssignment(
          node,
          node.modifiers,
          traversal.factory.updateObjectLiteralExpression(exportedExpression, properties),
        )
      }

      return traversal.factory.updateBinaryExpression(
        node,
        node.left,
        node.operatorToken,
        traversal.factory.updateObjectLiteralExpression(exportedExpression, properties),
      )
    }

    return node
  })

  /** config after transformations */
  const finalText = tailwindConfig.print()

  /**
   * No material changes could be made to the config,
   * so we warn the user and skip this step.
   *
   * This could be due to:
   * - the config structure being different than expected
   * - the config already being set up
   * - the config being empty
   * - the config file being invalid
   */
  if (initialText === finalText) {
    printManualTailwindSteps(
      cwd,
      subframeDirPath,
      "Subframe could not automatically configure your Tailwind config. To setup manually:",
    )
    return
  }

  tailwindConfig.formatText({
    indentSize: 2,
  })
}
