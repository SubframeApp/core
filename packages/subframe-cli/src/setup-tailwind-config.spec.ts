import { Project, SourceFile } from "ts-morph"
import { transformTailwindConfigFile } from "./setup-tailwind-config"

// Taken from Next13's default tailwind config
const DEFAULT_TAILWIND_CONFIG = {
  content: [
    "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}",
  ],
  theme: {
    extend: {
      backgroundImage: {
        "gradient-radial": "radial-gradient(var(--tw-gradient-stops))",
        "gradient-conic": "conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))",
      },
    },
  },
  plugins: [],
}

// Taken from Tailwind's Vite guide
// Modified to have a presets property
// https://tailwindcss.com/docs/guides/vite
const VITE_TAILWIND_CONFIG_WITH_PRESETS_PROPERTY = {
  content: ["./index.html", "./src/**/*.{js,ts,jsx,tsx}"],
  theme: {
    extend: {},
  },
  plugins: [],
  presets: [],
}

const CWD = "cwd/path"
const SUBFRAME_DIR_PATH = "subframe-dir/path"

function makeTailwindConfigContents(config: object, useESM: boolean) {
  if (useESM) {
    return `/** @type {import('tailwindcss').Config} */
export default ${JSON.stringify(config, null, 2)}`
  }
  return `/** @type {import('tailwindcss').Config} */
  module.exports = ${JSON.stringify(config, null, 2)}
  `
}

function getTransformedTailwindConfigFile(config: object, useESM = false): [SourceFile, SourceFile] {
  const project = new Project({ compilerOptions: { allowJs: true } })

  // Taken from Next13's default tailwind config
  const contents = makeTailwindConfigContents(config, useESM)

  const beforeFile = project.createSourceFile("./junk/delete-me/before", contents)
  const afterFile = project.createSourceFile("./junk/delete-me/after", contents)
  transformTailwindConfigFile(afterFile, CWD, SUBFRAME_DIR_PATH)
  return [beforeFile, afterFile]
}

describe("#transformTailwindConfig", () => {
  it("adds content to empty NextJS tailwind config", () => {
    const [_before, after] = getTransformedTailwindConfigFile({})
    expect(after.getText()).toMatchSnapshot()
  })

  it("adds content to default NextJS tailwind config", () => {
    const [_before, after] = getTransformedTailwindConfigFile(DEFAULT_TAILWIND_CONFIG)
    expect(after.getText()).toMatchSnapshot()
  })

  it("adds content to default Vite tailwind config with presets property", () => {
    const [_before, after] = getTransformedTailwindConfigFile(VITE_TAILWIND_CONFIG_WITH_PRESETS_PROPERTY)
    expect(after.getText()).toMatchSnapshot()
  })

  it("adds content to ESM-style file", () => {
    const [_before, after] = getTransformedTailwindConfigFile(DEFAULT_TAILWIND_CONFIG, true)
    expect(after.getText()).toMatchSnapshot()
  })

  it("running it the second time does not change the file", () => {
    const [_before, after] = getTransformedTailwindConfigFile(DEFAULT_TAILWIND_CONFIG)
    const firstRun = after.getText()
    transformTailwindConfigFile(after, CWD, SUBFRAME_DIR_PATH)
    expect(after.getText()).toEqual(firstRun)
  })
})
