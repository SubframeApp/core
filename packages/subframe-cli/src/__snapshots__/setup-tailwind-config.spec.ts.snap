// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#transformTailwindConfig adds content to ESM-style file 1`] = `
"export default {
  "content": ["./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}", "./../../subframe-dir/path/**/*.{tsx,ts,js,jsx}"],
  "theme": {
    "extend": {
      "backgroundImage": {
        "gradient-radial": "radial-gradient(var(--tw-gradient-stops))",
        "gradient-conic": "conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))"
      }
    }
  },
  "plugins": [],
  presets: [require("./../../subframe-dir/path/tailwind.config.js")]
};
"
`;

exports[`#transformTailwindConfig adds content to default NextJS tailwind config 1`] = `
"module.exports = {
  "content": ["./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
    "./src/app/**/*.{js,ts,jsx,tsx,mdx}", "./../../subframe-dir/path/**/*.{tsx,ts,js,jsx}"],
  "theme": {
    "extend": {
      "backgroundImage": {
        "gradient-radial": "radial-gradient(var(--tw-gradient-stops))",
        "gradient-conic": "conic-gradient(from 180deg at 50% 50%, var(--tw-gradient-stops))"
      }
    }
  },
  "plugins": [],
  presets: [require("./../../subframe-dir/path/tailwind.config.js")]
}
"
`;

exports[`#transformTailwindConfig adds content to default Vite tailwind config with presets property 1`] = `
"module.exports = {
  "content": ["./index.html",
    "./src/**/*.{js,ts,jsx,tsx}", "./../../subframe-dir/path/**/*.{tsx,ts,js,jsx}"],
  "theme": {
    "extend": {}
  },
  "plugins": [],
  "presets": [require("./../../subframe-dir/path/tailwind.config.js")]
}
"
`;

exports[`#transformTailwindConfig adds content to empty NextJS tailwind config 1`] = `
"module.exports = { content: ["./../../subframe-dir/path/**/*.{tsx,ts,js,jsx}"], presets: [require("./../../subframe-dir/path/tailwind.config.js")] }
"
`;
